File: cliparr-next/backend/app/db/init_db.py
import sqlite3
import os
import logging
from pathlib import Path

def init_db(db_path='data/cliparr.db'):
    """
    Initialize the database with all necessary tables.
    
    :param db_path: Path to the SQLite database file
    """
    try:
        # Ensure the directory exists
        os.makedirs(os.path.dirname(db_path), exist_ok=True)

        # Connect to the database
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()

        # Create shows table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS shows (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT NOT NULL,
                sonarr_id INTEGER UNIQUE,
                path TEXT,
                overview TEXT,
                status TEXT,
                added_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')

        # Create seasons table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS seasons (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                show_id INTEGER,
                season_number INTEGER,
                monitored BOOLEAN DEFAULT 1,
                FOREIGN KEY (show_id) REFERENCES shows(id) ON DELETE CASCADE,
                UNIQUE(show_id, season_number)
            )
        ''')

        # Create episodes table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS episodes (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                season_id INTEGER,
                sonarr_episode_id INTEGER UNIQUE,
                episode_number INTEGER,
                title TEXT,
                monitored BOOLEAN DEFAULT 1,
                FOREIGN KEY (season_id) REFERENCES seasons(id) ON DELETE CASCADE
            )
        ''')

        # Create episode_files table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS episode_files (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                episode_id INTEGER,
                file_path TEXT,
                size INTEGER,
                quality TEXT,
                FOREIGN KEY (episode_id) REFERENCES episodes(id) ON DELETE CASCADE
            )
        ''')

        # Create settings table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS settings (
                key TEXT PRIMARY KEY,
                value TEXT
            )
        ''')

        # Create indexes for performance
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_shows_title ON shows(title COLLATE NOCASE)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_shows_sonarr_id ON shows(sonarr_id)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_seasons_show_id ON seasons(show_id)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_episodes_season_id ON episodes(season_id)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_episodes_sonarr_episode_id ON episodes(sonarr_episode_id)')

        # Commit changes
        conn.commit()

        # Log successful initialization
        logging.info(f"Database initialized successfully at {db_path}")

    except sqlite3.Error as e:
        logging.error(f"Error initializing database: {e}")
        raise
    finally:
        if conn:
            conn.close()
# Run initialization if this script is run directly
if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    init_db()

File: cliparr-node.old/src/backend/models/initDb.ts
import pino from 'pino';
import pool from './db';
import { db } from '../config/database';
import { logger } from '../utils/logger';

const loggerPg = pino({
  transport:
    process.env.NODE_ENV !== 'production'
      ? {
          target: 'pino-pretty',
          options: { colorize: true },
        }
      : undefined,
});

export async function initializeDatabase(): Promise<void> {
  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    // Create shows table
    await client.query(`
      CREATE TABLE IF NOT EXISTS shows (
        id SERIAL PRIMARY KEY,
        sonarr_id INTEGER UNIQUE NOT NULL,
        title TEXT NOT NULL,
        overview TEXT,
        path TEXT,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
      )
    `);

    // Create seasons table
    await client.query(`
      CREATE TABLE IF NOT EXISTS seasons (
        id SERIAL PRIMARY KEY,
        show_id INTEGER REFERENCES shows(id) ON DELETE CASCADE,
        season_number INTEGER NOT NULL,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        UNIQUE(show_id, season_number)
      )
    `);

    // Create episodes table
    await client.query(`
      CREATE TABLE IF NOT EXISTS episodes (
        id SERIAL PRIMARY KEY,
        season_id INTEGER REFERENCES seasons(id) ON DELETE CASCADE,
        episode_number INTEGER NOT NULL,
        title TEXT,
        sonarr_episode_id INTEGER UNIQUE NOT NULL,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
      )
    `);

    // Create episode_files table
    await client.query(`
      CREATE TABLE IF NOT EXISTS episode_files (
        id SERIAL PRIMARY KEY,
        episode_id INTEGER REFERENCES episodes(id) ON DELETE CASCADE,
        file_path TEXT NOT NULL,
        size BIGINT,
        quality TEXT,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        UNIQUE(episode_id)
      )
    `);

    // Create settings table
    await client.query(`
      CREATE TABLE IF NOT EXISTS settings (
        key TEXT PRIMARY KEY,
        value TEXT NOT NULL,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
      )
    `);

    // Create indexes for better query performance
    await client.query('CREATE INDEX IF NOT EXISTS idx_shows_sonarr_id ON shows(sonarr_id)');
    await client.query('CREATE INDEX IF NOT EXISTS idx_seasons_show_id ON seasons(show_id)');
    await client.query('CREATE INDEX IF NOT EXISTS idx_episodes_season_id ON episodes(season_id)');
    await client.query(
      'CREATE INDEX IF NOT EXISTS idx_episodes_sonarr_id ON episodes(sonarr_episode_id)'
    );
    await client.query(
      'CREATE INDEX IF NOT EXISTS idx_episode_files_episode_id ON episode_files(episode_id)'
    );

    // Insert default settings if they don't exist
    await client.query(`
      INSERT INTO settings (key, value)
      VALUES ('import_mode', 'none')
      ON CONFLICT (key) DO NOTHING
    `);

    await client.query('COMMIT');
    loggerPg.info('Database initialized successfully');
  } catch (error) {
    await client.query('ROLLBACK');
    loggerPg.error('Error initializing database:', error);
    throw error;
  } finally {
    client.release();
  }
}

// Function to check if database is initialized
export async function isDatabaseInitialized(): Promise<boolean> {
  try {
    const result = (await db
      .prepare(
        `
      SELECT name FROM sqlite_master 
      WHERE type='table' AND name='shows'
    `
      )
      .get()) as { name: string } | undefined;
    return result !== undefined;
  } catch (error) {
    logger.error('Error checking database initialization:', error);
    throw error;
  }
}
